migrate timeout in Go Integration Cases

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.


### Version

unstable

### Minimal reproduce step

the ref to CI jobs: https://github.com/infdahai/incubator-kvrocks/actions/runs/5032023303/jobs/9025509337

The change is on https://github.com/infdahai/incubator-kvrocks/compare/fb0e3d4...7a5a2c2. I only run 10  tests in a loop.

### What did you expect to see?

All tests pass， especially the `Migrate_sync_with_(or_without)_all_kinds_of_timeouts` test.

### What did you see instead?

  --- FAIL: TestSlotMigrateSync/MIGRATE_-_Migrate_sync_with_(or_without)_all_kinds_of_timeouts#09 (0.01s)
        slotmigrate_test.go:426: 
            	Error Trace:	/home/runner/work/incubator-kvrocks/incubator-kvrocks/tests/gocase/integration/slotmigrate/slotmigrate_test.go:426
            	Error:      	Not equal: 
            	            	expected: string("OK")
            	            	actual  : <nil>(<nil>)
            	Test:       	TestSlotMigrateSync/MIGRATE_-_Migrate_sync_with_(or_without)_all_kinds_of_timeouts#09

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/src/cluster/sync_migrate_context.cc
+++ b/src/cluster/sync_migrate_context.cc
@@ -26,8 +26,7 @@ void SyncMigrateContext::Suspend() {
 
   if (timeout_ > 0) {
     timer_.reset(NewTimer(bufferevent_get_base(bev)));
-    int sec = static_cast<int>(timeout_);
-    timeval tm = {sec, static_cast<int>((timeout_ - static_cast<float>(sec)) * 1000000)};
+    timeval tm = {timeout_, 0};
     evtimer_add(timer_.get(), &tm);
   }
 }

--- a/src/cluster/sync_migrate_context.h
+++ b/src/cluster/sync_migrate_context.h
@@ -25,7 +25,7 @@
 class SyncMigrateContext : private EvbufCallbackBase<SyncMigrateContext, false>,
                            private EventCallbackBase<SyncMigrateContext> {
  public:
-  SyncMigrateContext(Server *svr, redis::Connection *conn, float timeout) : svr_(svr), conn_(conn), timeout_(timeout){};
+  SyncMigrateContext(Server *svr, redis::Connection *conn, int timeout) : svr_(svr), conn_(conn), timeout_(timeout){};
 
   void Suspend();
   void Resume(const Status &migrate_result);
@@ -36,7 +36,7 @@ class SyncMigrateContext : private EvbufCallbackBase<SyncMigrateContext, false>,
  private:
   Server *svr_;
   redis::Connection *conn_;
-  float timeout_ = 0;
+  int timeout_ = 0;
   UniqueEvent timer_;
 
   Status migrate_result_;

--- a/src/commands/cmd_cluster.cc
+++ b/src/commands/cmd_cluster.cc
@@ -151,7 +151,7 @@ class CommandClusterX : public Commander {
               return {Status::RedisParseErr, "timeout is not an integer or out of range"};
             }
             if (*parse_result < 0) {
-              return {Status::RedisParseErr, "timeout should not be negative"};
+              return {Status::RedisParseErr, errTimeoutIsNegative};
             }
             sync_migrate_timeout_ = *parse_result;
           }

--- a/src/commands/cmd_cluster.cc
+++ b/src/commands/cmd_cluster.cc
@@ -146,7 +146,14 @@ class CommandClusterX : public Commander {
           sync_migrate_ = true;
 
           if (args.size() == 6) {
-            sync_migrate_timeout_ = GET_OR_RET(ParseFloat<float>(args[5]));
+            auto parse_result = ParseInt<int>(args[5], 10);
+            if (!parse_result) {
+              return {Status::RedisParseErr, "timeout is not an integer or out of range"};
+            }
+            if (*parse_result < 0) {
+              return {Status::RedisParseErr, "timeout should not be negative"};
+            }
+            sync_migrate_timeout_ = *parse_result;
           }
         } else {
           return {Status::RedisParseErr, "Invalid sync flag"};
@@ -277,7 +284,7 @@ class CommandClusterX : public Commander {
   bool force_ = false;
 
   bool sync_migrate_ = false;
-  float sync_migrate_timeout_ = 0;
+  int sync_migrate_timeout_ = 0;
   std::unique_ptr<SyncMigrateContext> sync_migrate_ctx_ = nullptr;
 };
 

--- a/tests/gocase/integration/slotmigrate/slotmigrate_test.go
+++ b/tests/gocase/integration/slotmigrate/slotmigrate_test.go
@@ -413,22 +413,18 @@ func TestSlotMigrateSync(t *testing.T) {
 
 	t.Run("MIGRATE - Migrate sync with (or without) all kinds of timeouts", func(t *testing.T) {
 		slot++
-		require.Equal(t, "OK", rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync").Val())
+		// migrate sync without explicitly specify the timeout
+		result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync")
+		require.NoError(t, result.Err())
+		require.Equal(t, "OK", result.Val())
 
-		slot++
-		require.Equal(t, "OK", rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", -1).Val())
-
-		slot++
-		require.Equal(t, "OK", rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", 0).Val())
-
-		slot++
-		require.Equal(t, "OK", rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", 10).Val())
-
-		slot++
-		require.Equal(t, "OK", rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", 0.5).Val())
-
-		slot++
-		require.Equal(t, "OK", rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", -3.14).Val())
+		timeouts := []interface{}{-1, 0, 10, 0.5, -3.14}
+		for _, timeout := range timeouts {
+			slot++
+			result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", timeout)
+			require.NoError(t, result.Err())
+			require.Equal(t, "OK", result.Val())
+		}
 	})
 
 	t.Run("MIGRATE - Migrate sync timeout", func(t *testing.T) {

--- a/tests/gocase/integration/slotmigrate/slotmigrate_test.go
+++ b/tests/gocase/integration/slotmigrate/slotmigrate_test.go
@@ -435,17 +435,15 @@ func TestSlotMigrateSync(t *testing.T) {
 	})
 
 	t.Run("MIGRATE - Migrate sync timeout", func(t *testing.T) {
-		cnt := 200000
 		slot++
+		cnt := 200000
 		for i := 0; i < cnt; i++ {
 			require.NoError(t, rdb0.LPush(ctx, util.SlotTable[slot], i).Err())
 		}
-		timeout := 0.001
 
-		require.Nil(t, rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", timeout).Val())
-
-		// check the following command on the same connection
-		require.Equal(t, "PONG", rdb0.Ping(ctx).Val())
+		timeout := 1
+		result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", timeout)
+		require.Nil(t, result.Val())
 	})
 }
 

--- a/tests/gocase/integration/slotmigrate/slotmigrate_test.go
+++ b/tests/gocase/integration/slotmigrate/slotmigrate_test.go
@@ -418,7 +418,7 @@ func TestSlotMigrateSync(t *testing.T) {
 		require.NoError(t, result.Err())
 		require.Equal(t, "OK", result.Val())
 
-		invalidTimeouts := []interface{}{-2, -1, "abc"}
+		invalidTimeouts := []interface{}{-2, -1, 1.2, "abc"}
 		for _, timeout := range invalidTimeouts {
 			slot++
 			result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", timeout)

--- a/tests/gocase/integration/slotmigrate/slotmigrate_test.go
+++ b/tests/gocase/integration/slotmigrate/slotmigrate_test.go
@@ -412,7 +412,7 @@ func TestSlotMigrateSync(t *testing.T) {
 	})
 
 	t.Run("MIGRATE - Migrate sync with (or without) all kinds of timeouts", func(t *testing.T) {
-		slot++
+		slot := 1024
 		// migrate sync without explicitly specify the timeout
 		result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync")
 		require.NoError(t, result.Err())

--- a/tests/gocase/integration/slotmigrate/slotmigrate_test.go
+++ b/tests/gocase/integration/slotmigrate/slotmigrate_test.go
@@ -418,12 +418,19 @@ func TestSlotMigrateSync(t *testing.T) {
 		require.NoError(t, result.Err())
 		require.Equal(t, "OK", result.Val())
 
-		timeouts := []interface{}{-1, 0, 20, 10.5, -3.14}
-		for _, timeout := range timeouts {
+		invalidTimeouts := []interface{}{-2, -1, "abc"}
+		for _, timeout := range invalidTimeouts {
+			slot++
+			result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", timeout)
+			require.Error(t, result.Err(), "with timeout: %v", timeout)
+		}
+
+		validTimeouts := []int{0, 10, 20, 30}
+		for _, timeout := range validTimeouts {
 			slot++
 			result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", timeout)
 			require.NoError(t, result.Err(), "with timeout: %v", timeout)
-			require.Equal(t, "OK", result.Val(), "with timeout: %v", timeout)
+			require.Equal(t, "OK", result.Val(), "with timeout: %d", timeout)
 		}
 	})
 

--- a/tests/gocase/integration/slotmigrate/slotmigrate_test.go
+++ b/tests/gocase/integration/slotmigrate/slotmigrate_test.go
@@ -412,7 +412,7 @@ func TestSlotMigrateSync(t *testing.T) {
 	})
 
 	t.Run("MIGRATE - Migrate sync with (or without) all kinds of timeouts", func(t *testing.T) {
-		slot := 1024
+		slot++
 		// migrate sync without explicitly specify the timeout
 		result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync")
 		require.NoError(t, result.Err())
@@ -422,7 +422,7 @@ func TestSlotMigrateSync(t *testing.T) {
 		for _, timeout := range timeouts {
 			slot++
 			result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", timeout)
-			require.NoError(t, result.Err())
+			require.NoError(t, result.Err(), "without timeout: %v", timeout)
 			require.Equal(t, "OK", result.Val())
 		}
 	})

--- a/tests/gocase/integration/slotmigrate/slotmigrate_test.go
+++ b/tests/gocase/integration/slotmigrate/slotmigrate_test.go
@@ -422,8 +422,8 @@ func TestSlotMigrateSync(t *testing.T) {
 		for _, timeout := range timeouts {
 			slot++
 			result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", timeout)
-			require.NoError(t, result.Err(), "without timeout: %v", timeout)
-			require.Equal(t, "OK", result.Val())
+			require.NoError(t, result.Err(), "with timeout: %v", timeout)
+			require.Equal(t, "OK", result.Val(), "with timeout: %v", timeout)
 		}
 	})
 

--- a/tests/gocase/integration/slotmigrate/slotmigrate_test.go
+++ b/tests/gocase/integration/slotmigrate/slotmigrate_test.go
@@ -418,7 +418,7 @@ func TestSlotMigrateSync(t *testing.T) {
 		require.NoError(t, result.Err())
 		require.Equal(t, "OK", result.Val())
 
-		timeouts := []interface{}{-1, 0, 10, 0.5, -3.14}
+		timeouts := []interface{}{-1, 0, 20, 10.5, -3.14}
 		for _, timeout := range timeouts {
 			slot++
 			result := rdb0.Do(ctx, "clusterx", "migrate", slot, id1, "sync", timeout)

